### PR TITLE
Make `Task::block` and `Task::unblock` fallible

### PR DESCRIPTION
**Pull request by tsoutsman**
Previously, exited tasks could be blocked or unblocked leading to a panic during context switching.

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

### CODE REVIEW

File: applications/shell/src/lib.rs
Comment (kevinaboos) [inline applications/shell/src/lib.rs]: no, if the unblock operation fails because the task is exited here, there we should mark it as such and remove the job.

File: applications/test_wait_queue/src/lib.rs
Comment (kevinaboos) [inline applications/test_wait_queue/src/lib.rs:57]: lol can't believe I left that in. Feel free to remove haha

File: kernel/serial_port/src/lib.rs
Comment (kevinaboos) [inline kernel/serial_port/src/lib.rs]: i find this kind of formatting really tedious and hard to read. Personally would prefer something like:
```rust
Box::new(move || deferred_task.unblock()
    .expect(...)
)
```

File: kernel/serial_port/src/lib.rs
Comment (kevinaboos) [inline kernel/serial_port/src/lib.rs]: same formatting comment as above

File: kernel/sleep/src/lib.rs
Comment (kevinaboos) [inline kernel/sleep/src/lib.rs]: see top-level review comment about forwarding `Result`s instead of `unwrap`ping. This however is probably the least good option since it just silently masks failures.

File: kernel/task/src/lib.rs
Comment (kevinaboos) [inline kernel/task/src/lib.rs]: 1. Should probably return the current runstate on error for extra info
2. Clarify the docs to state that the function only succeeds if the current runstate is `Runnable`
3. Clarify the docs to state that it returns the previous runstate on success and the current runstate on error

File: kernel/task/src/lib.rs
Comment (kevinaboos) [inline kernel/task/src/lib.rs]: IMO these should be re-ordered to move the common success case to the top, where the current runstate is `Runnable` and the new one is `Blocked`. That way we're not always wasting cycles on an expensive atomic conditional for a case (`Initing` --> `Blocked`) that will likely never occur.

I think it should be an error to attempt to block a task that is in the `Initing` runstate, since that means something went wrong with initialization and you were able to obtain a reference to a task that wasn't yet finished being initialized. I don't believe that could happen, but better safe than sorry.
For the one case where this is currently accepted (spawning a task that shouldn't run), perhaps we can add a separate code path to allow that transition in only this case.

Could you also condense the formatting a bit?

Comment (kevinaboos) [reply kernel/task/src/lib.rs]: Also should it be considered an error to attempt to block a task that is already blocked? Might be indicative of a logic error in the calling application; currently it's fine and doesn't hurt, but still worth raising the question.

File: kernel/task/src/lib.rs
Comment (kevinaboos) [inline kernel/task/src/lib.rs]: same comments as for `block()` above

File: kernel/wait_queue/src/lib.rs
Comment (kevinaboos) [inline kernel/wait_queue/src/lib.rs:209]: i think this could cause an infinite loop; any options to avoid that problem?

**Review by kevinaboos — CHANGES_REQUESTED** [review]
Thanks for the PR, I agree this can help avoid potential weird task behavior if someone is able to get ahold of an exited task and mark it as unblocked. 

I left a few comments; aside from that, I think it'd be nice to return `Result`s wherever possible instead of panicking on an `unwrap`. If there are places where you're unsure how best to bubble up those `Result`s, just leave a comment denoting where and I can handle them.

Thoughts for a future PR: eventually I'd like to completely redesign task blocking to have some kind of "`BlockGuard`", with the idea being that when a task `A` has blocked another task `B`, if task `A` crashes and panics we can then drop that `BlockGuard` during unwinding, which will then unblock task `B`. Right now without that feature, task `B` will never be unblocked if task `A` fails, which is sort of a less-than-ideal outcome, at least from the mindset of fault tolerance.

File: kernel/wait_queue/src/lib.rs
Comment (tsoutsman) [reply kernel/wait_queue/src/lib.rs:209]: I don't think it could because on every loop an item gets removed from `wq_locked` and so eventually `wq_locked` will run out of items and the function will return `false`.

File: kernel/serial_port/src/lib.rs
Comment (tsoutsman) [reply kernel/serial_port/src/lib.rs]: The best I can do is [this](https://github.com/theseus-os/Theseus/pull/657/commits/00880ec54bdd6e04ac021ce41524c8fe1cab34af). Your suggestion won't work because if the closure is more than one line it must be delimited using curly braces.

File: kernel/sleep/src/lib.rs
Comment (tsoutsman) [reply kernel/sleep/src/lib.rs]: should we log something here? is a task being killed while it is sleeping a failure?

File: kernel/task/src/lib.rs
Comment (tsoutsman) [reply kernel/task/src/lib.rs]: By separate code path do you mean a function?

> Also should it be considered an error to attempt to block a task that is already blocked?

Maybe? The only legitimate use case I can currently think of is the shell processing a Ctrl+Z signal while running a task that is already blocked by something (e.g. Mutex). But in my rework of the `tty`, I add a separate `Suspended` task state similar to Linux's `STOPPED` which ensures the two states aren't mixed.

File: applications/test_wait_queue/src/lib.rs
Comment (tsoutsman) [reply applications/test_wait_queue/src/lib.rs:57]: I like it. It's like a historical artifcat.

File: kernel/task/src/lib.rs
Comment (kevinaboos) [reply kernel/task/src/lib.rs]: > By separate code path do you mean a function?

yes, sorry, i meant exposing a new dedicated function within `task` for this purpose alone, i.e., something like `block_new_task()` or `block_initing_task()` which only allows a Runstate transition from `Initing` --> `Blocked`.

Comment (kevinaboos) [reply kernel/task/src/lib.rs]: As for blocking an already-blocked task, for now let's just log a warning and see where/when it occurs.

File: kernel/wait_queue/src/lib.rs
Comment (kevinaboos) [reply kernel/wait_queue/src/lib.rs:209]: good enough for me :smile:

File: kernel/serial_port/src/lib.rs
Comment (kevinaboos) [reply kernel/serial_port/src/lib.rs]: ah i see. It's not because it's multiple lines, it's because we don't wan't to return the value of the last expression in the block, we want the return to be void `()`.

anyway, good enough.

File: kernel/sleep/src/lib.rs
Comment (kevinaboos) [reply kernel/sleep/src/lib.rs]: yep, it's worth logging and then unwrapping. The whole `sleep` crate will be redesigned once we have better sync primitives, e.g., the ones you're working on as part of `std` stuff.

Another eventual goal is to force a killed task to unwind (and thus, release its resources) rather than just stopping its execution immediately. That'd be relevant here too, but for now I think an `unwrap` call is warranted.

**Comment by tsoutsman** [comment]
Should I remove the unwraps from the test applications as well?

Also FYI, I originally encountered this issue when killing a task waiting on a mutex. The task would remain in the wait queue and hence be unblocked.

**Comment by kevinaboos** [comment]
> Should I remove the unwraps from the test applications as well?

If you want, but IMO it's fine if application code panics.

> Also FYI, I originally encountered this issue when killing a task waiting on a mutex. The task would remain in the wait queue and hence be unblocked.

I see. That sort of odd behavior is also a good motivating factor for why we need the above `BlockGuard`-esque solution.

**Comment by kevinaboos** [comment]
Looks good as of now, I made a few small changes and am testing some things.

@tsoutsman btw, when running this branch did you ever encounter any of the new `warn!()` log statements that you added about blocked tasks being blocked again or unblocked tasks being unblocked again? I wouldn't want to flood the log with warnings if that's common behavior (but afaik i try to avoid doing spurious wakeups like that).

**Review by kevinaboos — APPROVED** [review]
I made a few small changes and removed the one case of a blocked task being re-blocked (in the shell). Ready to merge, thanks!

**Comment by tsoutsman** [comment]
> @tsoutsman btw, when running this branch did you ever encounter any of the new warn!() log statements that you added about blocked tasks being blocked again or unblocked tasks being unblocked again?

yea the shell doesn't yield after blocking itself leading to a couple of logs at startup. I've fixed it in #675 but the scope of that PR is much larger.

**Pull request merged**
